lib/enkf/row_scaling.cpp

```diff
@@ -21,12 +21,152 @@
+
+#define ROW_SCALING_ID 6123199
+#define ROW_SCALING_RESOLUTION 1000
+
+
+/*
+  The values in the row_scaling container are distributed among
+  ROW_SCALING_RESOLUTION discreete values. The reason for this lumping is to be
+  able to reuse the scaled update matrix:
+
+      X(a) = a*X0 + (1 - a)*I
+
+  for several rows. This grouping is utilized in the row_scaling_multiply()
+  function.
+*/
+
+
+namespace {
+
+void scaleX(matrix_type* X, const matrix_type * X0, double alpha) {
+  matrix_assign(X, X0);
+  matrix_scale(X, alpha);
+  for (int i=0; i < matrix_get_rows(X); i++)
+    matrix_iset(X, i, i, (1 - alpha) + matrix_iget(X, i, i));
+}
+
+}
+
+int row_scaling::size() const {
+  return this->data.size();
+}
+
+
+double row_scaling::operator[](int index) const {
+  return this->data.at(index);
+}
+
+void row_scaling::assign(int index, double value) {
+  if (value < 0 || value > 1)
+    throw std::invalid_argument("Invalid value ");
+
+  if (this->data.size() <= index)
+    this->data.resize(index + 1, 1);
+
+  this->data.at(index) = floor(value * this->resolution) / this->resolution;
+}
+
+void row_scaling::multiply(matrix_type * A, const matrix_type * X0) const {
+  if (this->data.size() != matrix_get_rows(A))
+    throw std::invalid_argument("Size mismatch between row_scaling and A matrix");
+
+  if (matrix_get_columns(A) != matrix_get_rows(X0))
+    throw std::invalid_argument("Size mismatch between X0 and A matrix");
+
+  if (matrix_get_columns(X0) != matrix_get_rows(X0))
+    throw std::invalid_argument("X0 matrix is not quadratic");
+
+  matrix_type * X = matrix_alloc(matrix_get_rows(X0), matrix_get_columns(X0));
+  std::vector<int> sort_index(this->size());
+  std::iota(sort_index.begin(), sort_index.end(), 0);
+  std::sort(sort_index.begin(), sort_index.end(), [this] (int index1, int index2) { return this->operator[](index1) > this->operator[](index2); });
+
+  std::size_t index_offset = 0;
+  while (true) {
+    auto end_index = index_offset;
+    auto current_alpha = this->data[sort_index[end_index]];
+    std::vector<int> row_list;
+
+    while (true) {
+      if (end_index == this->data.size())
+        break;
+
+      if (this->data[sort_index[end_index]] != current_alpha )
+        break;
+
+      row_list.push_back(sort_index[end_index]);
+      end_index += 1;
+    }
+    if (row_list.empty())
+      break;
+
+    double alpha = this->data[row_list[0]];
+    if (alpha == 0.0)
+      break;
+
+    scaleX(X, X0, alpha);
+    const double * X_data = matrix_get_data(X);
+
+    for (const auto& row : row_list) {
+      std::vector<double> src_row(matrix_get_columns(A));
+      std::vector<double> target_row(matrix_get_columns(A));
+
+      for (int j=0; j < matrix_get_columns(A); j++)
+        src_row[j] = matrix_iget(A, row, j);
+
+      for (int j=0; j < matrix_get_columns(A); j++) {
+        double sum = 0;
+        for (int i=0; i < matrix_get_columns(A); i++)
+          sum += src_row[i] * X_data[GET_INDEX(X, i, j)];
```

Comment:
Why are we using the one-dimensional representation of X

Reply:
For efficiency; I can change that.

Reply:
no need... was just curious